sdk/storage/azure-storage-file-share/src/main/java/com/azure/storage/file/share/StorageSeekableByteChannelShareFileReadBehavior.java

```diff
@@ -0,0 +1,84 @@
+import com.azure.storage.file.share.options.ShareFileDownloadOptions;
+import com.fasterxml.jackson.databind.util.ByteBufferBackedOutputStream;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+class StorageSeekableByteChannelShareFileReadBehavior implements StorageSeekableByteChannel.ReadBehavior {
+    private static final ClientLogger LOGGER = new ClientLogger(StorageSeekableByteChannelShareFileReadBehavior.class);
+
+    private final ShareFileClient client;
+    private final ShareRequestConditions conditions;
+
+    private Long lastKnownResourceLength;
+
+    StorageSeekableByteChannelShareFileReadBehavior(ShareFileClient client, ShareRequestConditions conditions) {
+        this.client = client;
+        this.conditions = conditions;
+    }
+
+    ShareFileClient getClient() {
+        return this.client;
+    }
+
+    ShareRequestConditions getRequestConditions() {
+        return this.conditions;
+    }
+
+    @Override
+    public int read(ByteBuffer dst, long sourceOffset) {
+        if (dst.remaining() <= 0) {
+            throw LOGGER.logExceptionAsError(new IllegalArgumentException("'dst.remaining()' must be positive."));
+        }
+
+        int initialPosition = dst.position();
+
+        try (ByteBufferBackedOutputStream dstStream = new ByteBufferBackedOutputStream(dst)) {
+            ShareFileDownloadResponse response =  client.downloadWithResponse(dstStream,
+                new ShareFileDownloadOptions()
+                    .setRange(new ShareFileRange(sourceOffset, sourceOffset + dst.remaining() - 1))
+                    .setRequestConditions(conditions),
+                null, null);
+            lastKnownResourceLength = getResourceLengthFromContentRange(
+                response.getDeserializedHeaders().getContentRange());
+            return dst.position() - initialPosition;
+        } catch (ShareStorageException e) {
+            if (e.getErrorCode() == ShareErrorCode.INVALID_RANGE) {
+                String contentRange = e.getResponse().getHeaderValue("Content-Range");
+                if (contentRange != null) {
+                    lastKnownResourceLength = getResourceLengthFromContentRange(contentRange);
+                }
+                // if requested offset is past updated end of file, then signal end of file. Otherwise, only signal
+                // that zero bytes were read
+                return sourceOffset < lastKnownResourceLength ? 0 : -1;
+            }
+            throw LOGGER.logExceptionAsError(e);
+        } catch (IOException e) {
+            throw LOGGER.logExceptionAsError(new RuntimeException(e));
+        }
+    }
+
+    @Override
+    public long getResourceLength() {
+        if (lastKnownResourceLength == null) {
+            lastKnownResourceLength = client.getProperties().getContentLength();
+        }
+        return lastKnownResourceLength;
+    }
+
+    private static long getResourceLengthFromContentRange(String contentRange) {
```

Comment:
An API for this was added to `CoreUtils`, `CoreUtils.extractSizeFromContentRange`